### PR TITLE
Use recommended algorithm in assertions

### PR DESCRIPTION
**Pull request by Avery-Dunn**
Fixes the issue described in https://github.com/AzureAD/microsoft-authentication-library-for-java/issues/958

Microsoft recommends PS256 for assertions and other MSALs follow this, however MSAL Java still used RS256. It seems like PSS support was only added in Java 11, but it was thankfully backported to Java 8.

This PR refactors JwtHelper to use the recommended algorithm and adjusts/expands tests in order to cover the new behavior.

In addition, while making the changes I discovered a weird edge case:
- To create a Signature instance to sign the assertion, we need to provide a `PrivateKey`
- When retrieving a `PrivateKey` from the Windows-MY keystore the underlying type is "sun.security.mscapi.CPrivateKey"
- The `sun.security.mscapi.CPrivateKey` type was accepted by the Signature instance created with "SHA256withRSA", but for some reason the Signature created with "RSASSA-PSS" throws an error about it not specifically being an `RSAPrivateKey` instance

Since our integration tests use a cert stored in Windows-MY simply changing the assertion to use PS256 would've broken many of our tests. I'm sure plenty of our customers retrieve certs in the same way, and I'm not sure what other key types would cause this error.

So to avoid any breaking changes, this PR also implements a mechanism to fallback to the old RS256 style when we get a certain exception, and has extra tests covering this fallback behavior.

### CODE REVIEW

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/JwtHelper.java [hunk 1]
```diff
@@ -22,62 +25,183 @@ static ClientAssertion buildJwt(String clientId, final ClientCertificate credent
...
+     *
+     * @param clientId     The client ID to use as the issuer and subject
+     * @param credential   The certificate credential used for signing
+     * @param jwtAudience  The audience claim for the JWT
+     * @param sendX5c      Whether to include the x5c header with certificate chain
+     * @param useSha1      Whether to use SHA-1 hash for thumbprint instead of SHA-256
+     * @return A ClientAssertion containing the signed JWT
+     * @throws Exception If JWT creation or signing fails
+     */
+    private static ClientAssertion generateRs256Jwt(String clientId, ClientCertificate credential,
```
Comment (bgavrilMS) [inline msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/JwtHelper.java:92]: There seem to be a lot of code dup between thjis method and the ps256 version

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/JwtHelper.java [hunk 2]
```diff
@@ -22,62 +25,183 @@ static ClientAssertion buildJwt(String clientId, final ClientCertificate credent
...
-            // is not supported or the IClientCredential.publicCertificateHash256() method is not implemented, the library will default to SHA-1.
-            String hash256 = credential.publicCertificateHash256();
-            if (useSha1 || hash256 == null) {
-                header.put("x5t", credential.publicCertificateHash());
-            } else {
-                header.put("x5t#S256", hash256);
+            // First try with PS256 (preferred)
+            return generatePS256Jwt(clientId, credential, jwtAudience, sendX5c, useSha1);
+        } catch (InvalidKeyException e) {
+            // If the key isn't compatible with PSS, fall back to RS256.
```
Comment (bgavrilMS) [inline msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/JwtHelper.java:31]: This is very strange, because the padding algorithm can be always be used with a PSS key

image

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/HelperAndUtilityTests.java [hunk 3]
```diff
@@ -187,4 +190,99 @@ void JsonHelper_createIdTokenFromEncodedTokenString_InvalidJsonInToken() {
 
         assertEquals(AuthenticationErrorCode.INVALID_JSON, exception.errorCode());
     }
+
+    @Test
+    void JwtHelper_buildJwt_UsesPSS256WhenSupported() throws Exception {
```
Comment (bgavrilMS) [inline msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/HelperAndUtilityTests.java:195]: The way it's implemented in MSAL .NET is: 

- for AAD, use SHA2 and PSS
- for ADFS, use SHA1 and PKCS1


I see that the way to specify a certificate in MSAL Java is similar. So why not have some higher level tests, i.e. start from the public API and assert what gets put on the wire?

**Review by bgavrilMS — CHANGES_REQUESTED** [review]
RSA padding should be universal, but I understand that Java8 uses CAPI instead of CNG.

However, for backwards compat, in places where we use SHA1, we should continue to use PKCS1. Does this happen? Can we have better E2E tests?

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/JwtHelper.java (hunk 2 above)
Comment (bgavrilMS) [reply msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/JwtHelper.java:31]: Read more about this, CNG isn't available on Java 8.

**Comment by Avery-Dunn** [comment]
Closing, will need to investigate cryptography options more.
